src/operations/SymbolicOperation.cpp

```diff
@@ -0,0 +1,286 @@
+    }
+
+    SymbolicOperation::SymbolicOperation(dd::QubitCount nq, const dd::Controls& controls, dd::Qubit target, OpType g, const SymbolOrNumber& lambda, const SymbolOrNumber& phi, const SymbolOrNumber& theta, dd::Qubit startingQubit):
+        SymbolicOperation(nq, target, g, lambda, phi, theta, startingQubit) {
+        this->controls = controls;
+    }
+
+    SymbolicOperation::SymbolicOperation(dd::QubitCount nq, const dd::Controls& controls, const Targets& targets, OpType g, const SymbolOrNumber& lambda, const SymbolOrNumber& phi, const SymbolOrNumber& theta, dd::Qubit startingQubit):
+        SymbolicOperation(nq, targets, g, lambda, phi, theta, startingQubit) {
+        this->controls = controls;
+    }
+
+    // MCF (cSWAP), Peres, paramterized two target Constructor
+    SymbolicOperation::SymbolicOperation(dd::QubitCount nq, const dd::Controls& controls, dd::Qubit target0, dd::Qubit target1, OpType g, const SymbolOrNumber& lambda, const SymbolOrNumber& phi, const SymbolOrNumber& theta, dd::Qubit startingQubit):
+        SymbolicOperation(nq, controls, {target0, target1}, g, lambda, phi, theta, startingQubit) {
+    }
+
+    bool SymbolicOperation::equals(const Operation& op, const Permutation& perm1, const Permutation& perm2) const {
+        if (!op.isSymbolicOperation() && !isStandardOperation()) {
+            return false;
+        }
+        if (isStandardOperation() && qc::StandardOperation::equals(op, perm1, perm2)) return true;
+
+        if (!op.isSymbolicOperation()) return false;
+        const auto& symOp = dynamic_cast<const SymbolicOperation&>(op);
+        for (std::size_t i = 0; i < symbolicParameter.size(); ++i) {
+            if (symbolicParameter[i].has_value() != symOp.symbolicParameter[i].has_value())
+                return false;
+
+            if (symbolicParameter[i].has_value())
+                return symbolicParameter[i].value() == symOp.symbolicParameter[i].value();
+        }
+        return true;
+    }
+
+    void SymbolicOperation::dumpOpenQASM([[maybe_unused]] std::ostream& of, [[maybe_unused]] const RegisterNames& qreg, [[maybe_unused]] const RegisterNames& creg) const {
+        throw QFRException("OpenQasm2.0 doesn't support parametrized gates!");
+    }
+    void SymbolicOperation::dumpQiskit([[maybe_unused]] std::ostream& of, [[maybe_unused]] const RegisterNames& qreg, [[maybe_unused]] const RegisterNames& creg, [[maybe_unused]] const char* anc_reg_name) const {
+        throw QFRException("Dumping Qiskit Circuit is not supported for parameterized gates.");
+    }
+
+    StandardOperation SymbolicOperation::getInstantiatedOperation(const VariableAssignment& assignment) const {
+        auto lambda = getInstantiation(getParameter(0), assignment);
+        auto phi    = getInstantiation(getParameter(1), assignment);
+        auto theta  = getInstantiation(getParameter(2), assignment);
+        return StandardOperation(nqubits, targets, type, lambda, phi, theta, startQubit);
```

Comment:
while not as easy as the function below, this function could also benefit from a little more generality.
If this turns out to be harder, feel free to leave it as it is.

Reply:
That one I can't, right? The StandardOperation is not generic and has a set number of parameters. So it wouldn't make sense to construct it with a generic number of arguments. Correct me if my thinking is wrong on this one.

Reply:
The Operation class has a compile-time-set number of parameters. However, that number might not be `3` in the near to mid-term future (there is a global `U` gate in Qiskit that takes `4` parameters).
Due to the constructor taking a fixed number of arguments, it might actually be tough to make this generic. Could fold expressions be used here to expand the list of parameters?

Reply:
First comment in this thread seems to solve the issue: https://stackoverflow.com/questions/60434033/how-do-i-expand-a-compile-time-stdarray-into-a-parameter-pack . However, that is quite some template boilerplate :smile:

Reply:
Oh boy. Although I like fancy C++ concepts, that might be too much for such a simple use case.
Probably best to leave it as it is and live with the fact that this potentially has to be adapted in the future.